core/src/main/scala/kanaloa/reactive/dispatcher/queue/Worker.scala

```diff
@@ -7,186 +7,171 @@ import kanaloa.reactive.dispatcher.ApiProtocol.{QueryStatus, WorkFailed, WorkTim
-        case (workMsg, workSender) ⇒ workSender ! Rejected(workMsg, "retiring") //this happens only when a Retire is sent during a restart with new routee
-      }
-      finish()
+    //if there is no work left, or if the Queue dies, the Actor must wait for the Work to finish before terminating
+    case Terminated(`queue`) ⇒ context become waitingToTerminate(outstanding)
+    case NoWorkLeft          ⇒ context become waitingToTerminate(outstanding)
 
+    //This is a fun state.  The Worker is told to stop, but needs to both wait for Unregister and for Work to complete
+    case Worker.Retire ⇒
+      queue ! Unregister
+      context become unregisteringOutstanding(outstanding)
   }
 
-  val waitingForWork: Receive =
-    whileWaiting(Idle) orElse {
+  //This state waits for Work to complete, and then stops the Actor
+  def waitingToTerminate(outstanding: Outstanding): Receive = routeeResponse(outstanding, finish) orElse {
+    case qs: QueryStatus     ⇒ qs reply WaitingToTerminate
 
-      case work: Work ⇒ sendWorkToDelegatee(work, 0)
+    //ignore these, since all we care about is the Work completing one way or another
+    case Hold(x)             ⇒
+    case Retire              ⇒
+    case Terminated(`queue`) ⇒
+    case NoWorkLeft          ⇒
 
-      case Terminated(r) if r == routee ⇒
-        context become waitingForRoutee(None)
-        retrieveRoutee()
+    case w: Work             ⇒ sender() ! Rejected(w, "Retiring") //safety first
 
-      case Worker.Retire ⇒
-        queue ! Unregister(self)
-        context become retiring(None)
-    }
+    case WorkFinished        ⇒ finish() //work is done, terminate
+  }
 
-  def whileWaiting(currentStatus: WorkerStatus): Receive = {
-    case NoWorkLeft          ⇒ finish()
+  //in this state, we have told the Queue to Unregister this Worker, so we are waiting for an acknowledgement
+  def unregistering: Receive = {
+    case qs: QueryStatus      ⇒ qs reply Unregistering
 
-    case Hold(period)        ⇒ delayBeforeNextWork = Some(period)
+    //ignore these
+    case Hold(x)              ⇒
+    case Retire               ⇒
+    case Terminated(`routee`) ⇒
+    case NoWorkLeft           ⇒
 
-    case qs: QueryStatus     ⇒ qs reply currentStatus
+    case w: Work              ⇒ sender ! Rejected(w, "Retiring") //safety first
 
-    case Terminated(`queue`) ⇒ finish()
+    //Either we Unregistered successfully, or the Queue died.  terminate
+    case Unregistered         ⇒ finish()
+    case Terminated(`queue`)  ⇒ finish()
 
   }
 
-  def finish(): Unit = context stop self
-
-  def working(outstanding: Outstanding): Receive = ({
-    case Hold(period)        ⇒ delayBeforeNextWork = Some(period)
+  //in this state we are we waiting for 2 things to happen, Unregistration and Work completing
+  //the Worker will shift its state based on which one happens first
+  def unregisteringOutstanding(outstanding: Outstanding): Receive = routeeResponse(outstanding, becomeUnregistering) orElse {
```

Comment:
Actually, not sure how safe that it is.  I really wanted to make sure that Workers were unregistered before they terminate, just to avoid any potential work loss.  So, I don't think that one extra unregister message would cause any issues.